File: freeq-ios/freeq/Models/RowSignatureMark.swift
import Foundation

/// The signature mark a message row wears, decided from the row's verdict.
/// Pure/Foundation → unit-testable; the views only apply it. Mirrors the
/// macOS `RowSignatureMark`.
enum RowSignatureMark: Equatable {
    /// The sender's own device signed: a lock at `opacity`, full when the key
    /// is published in their account, 30% while only their server vouches.
    case lock(opacity: Double)
    /// The signature failed, or was made after its key was retired.
    case warning

    /// The lock's opacity while only the sender's server vouches for the key.
    static let vouchedOpacity = 0.3

    /// The mark for a verdict; nil with no verdict and for every state that
    /// shows nothing: server-signed, unsigned, unverifiable, still checking.
    static func of(_ verdict: VerdictInfo?) -> RowSignatureMark? {
        guard let verdict else { return nil }
        switch verdict.kind {
        case .device:
            return .lock(opacity: verdict.layer == .published ? 1 : vouchedOpacity)
        case .invalid, .retired:
            return .warning
        case .server, .unsigned, .unverifiable, .pending:
            return nil
        }
    }

    /// A row's mark once its check has answered; `mark` is nil for a verdict
    /// that shows nothing.
    struct Settled: Equatable {
        let mark: RowSignatureMark?
    }

    /// The settled mark for a verdict; nil with no verdict or while pending.
    static func settled(_ verdict: VerdictInfo?) -> Settled? {
        guard let verdict, verdict.kind != .pending else { return nil }
        return Settled(mark: of(verdict))
    }

    /// Whether a row that would group under its header starts its own header
    /// instead: only when both marks have settled and differ. Nil is pending,
    /// and a pending mark matches.
    static func startsHeader(header: Settled?, row: Settled?) -> Bool {
        guard let header, let row else { return false }
        return header != row
    }

    /// Which rows start a header, in one pass over the transcript. A row
    /// starts one when it breaks the sender's run, or when its settled mark
    /// differs from its header's, the nearest header above it in the run.
    static func headers(breaksRun: [Bool], marks: [Settled?]) -> [Bool] {
        var headers: [Bool] = []
        headers.reserveCapacity(breaksRun.count)
        var headerMark: Settled? = nil
        for (breaks, mark) in zip(breaksRun, marks) {
            let starts = breaks || startsHeader(header: headerMark, row: mark)
            if starts { headerMark = mark }
            headers.append(starts)
        }
        return headers
    }
}

File: freeq-android/freeq/src/main/java/com/freeq/model/RowSignatureMark.kt
package com.freeq.model

import com.freeq.ffi.KeyLayer
import com.freeq.ffi.VerdictState

/**
 * The signature mark a message row wears, decided from the row's verdict.
 *
 * Lives outside the view so the rule can be unit-tested without an Android
 * runtime; the row only applies it. Mirrors the iOS and macOS
 * `RowSignatureMark`.
 */
internal object RowSignatureMark {
    sealed interface Mark

    /** The sender's own device signed: a lock at [alpha], full when the key is
     *  published in their account, 30% while only their server vouches. */
    data class Lock(val alpha: Float) : Mark

    /** The signature failed, or was made after its key was retired. */
    object Warning : Mark

    /** The lock's alpha while only the sender's server vouches for the key. */
    const val VOUCHED_ALPHA = 0.3f

    /** The mark for a verdict; null with no verdict and for every state that
     *  shows nothing: server-signed, unsigned, unverifiable, still checking. */
    fun of(verdict: FfiVerdict?): Mark? {
        if (verdict == null) return null
        return when (verdict.state) {
            VerdictState.DEVICE ->
                Lock(if (verdict.layer == KeyLayer.PUBLISHED) 1f else VOUCHED_ALPHA)
            VerdictState.INVALID, VerdictState.RETIRED -> Warning
            VerdictState.SERVER, VerdictState.UNSIGNED,
            VerdictState.UNVERIFIABLE, VerdictState.PENDING -> null
        }
    }

    /** A row's mark once its check has answered; [mark] is null for a verdict
     *  that shows nothing. */
    data class Settled(val mark: Mark?)

    /** The settled mark for a verdict; null with no verdict or while pending. */
    fun settled(verdict: FfiVerdict?): Settled? =
        if (verdict == null || verdict.state == VerdictState.PENDING) null else Settled(of(verdict))

    /** Whether a row that would group under its header starts its own header
     *  instead: only when both marks have settled and differ. Null is pending,
     *  and a pending mark matches. */
    fun startsHeader(header: Settled?, row: Settled?): Boolean =
        header != null && row != null && header != row
}
